modules/internal/ChapelLocale-chpldoc.chpl

```diff
@@ -0,0 +1,608 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  A *locale* is a Chapel abstraction for a piece of a target
+  architecture that has processing and storage capabilities.
+  Generally speaking, the tasks running within a locale have
+  roughly uniform access to values stored in the locale's local
+  memory and longer latencies for accessing the memories of other
+  locales.  As examples, a single shared memory machine would be
+  defined as a single locale, while in a system consisting of a
+  group of network-connected multicore nodes or SMPs each node
+  would be defined as a locale.
+
+  Chapel provides several predefined methods on locales, as well as
+  a few variables that describe the locales upon which a program is
+  running.
+ */
+module ChapelLocale {
+
+  use LocaleModel;
+
+  //
+  // Node and sublocale types and special sublocale values.
+  //
+  pragma "no doc"
+  type chpl_nodeID_t = int(32);
+  pragma "no doc"
+  type chpl_sublocID_t = int(32);
+
+  pragma "no doc"
+  extern const c_sublocid_none: chpl_sublocID_t;
+  pragma "no doc"
+  extern const c_sublocid_any: chpl_sublocID_t;
+
+  /*
+    ``locale`` is the abstract class from which the various
+    implementations inherit.  It specifies the required interface
+    and implements part of it, but requires the rest to be provided
+    by the corresponding concrete classes.
+   */
+  class locale {
+    //- Constructor
+    pragma "no doc"
+    proc locale() { }
+
+    //------------------------------------------------------------------------{
+    //- Fields and accessors defined for all locale types (not overridable)
+    //-
+
+    // Every locale has a parent, except for the root locale.
+    // The parent of the root locale is nil (by definition).
+    pragma "no doc"
+    const parent : locale;
+
+    // To be removed from the required interface once legacy code is adjusted.
+    // Modified in RootLocale.init().
+
+    /*
+      This is the number of processors on this locale.  The term
+      *cores* here is a misnomer.  The value is actually the number
+      of instances of the CPU architecture, so for a locale with a
+      multithreaded hardware architecture, it's the number of
+      threads.  Also, only processors actually available to the
+      program are in this count.  If the OS has made some of them
+      off-limits to the program, for example, those are not
+      included.
+     */
+    var numCores: int;
+
+    /*
+      This is the maximum task concurrency that one can expect to
+      achieve on this locale.  This is often equal to `numCores`.
+      Creating more tasks than this will probably increase walltime
+      rather than decrease it.
+     */
+    var maxTaskPar: int; // max parallelism tasking layer expects to deliver
```

Comment:
What do you think of adding `proc`s with the same name as maxTaskPar, numCores, and callStackSize to match the format in the spec, and to avoid people from thinking about the inner details of the locale, and instead just it's interface?